dcos_launch/terraform.py

```diff
@@ -0,0 +1,100 @@
+        self.init_dir = os.path.abspath(self.init_dir)
+        self.cluster_profile_path = os.path.join(self.init_dir, 'desired_cluster_profile.tfvars')
+        if env is None:
+            env = os.environ.copy()
+
+        if config['platform'] == 'azure':
+            # TODO
+            pass
+        elif config['platform'] == 'aws':
+            # TODO
+            pass
+        elif config['platform'] == 'gcp':
+            if 'GCE_CREDENTIALS' in env:
+                with open('gcp_service_creds.json', 'w') as file:
+                    file.write(env['GCE_CREDENTIALS'])
+                config['terraform_config']['gcp_credentials_key_file'] = os.path.abspath('gcp_service_creds.json')
+            elif 'GOOGLE_APPLICATION_CREDENTIALS' in env:
+                config['terraform_config']['gcp_credentials_key_file'] = env['GOOGLE_APPLICATION_CREDENTIALS']
+            else:
+                raise util.LauncherError(
+                    'MissingParameter', 'Either GCE_CREDENTIALS or GOOGLE_APPLICATION_CREDENTIALS must be set in env')
+
+            if 'gcp_project' not in config['terraform_config']:
+                config['terraform_config']['gcp_project'] = json.load(
+                    config['terraform_config']['gcp_credentials_key_file'])['project_id']
+        else:
+            raise util.LauncherError('InvalidPlatform', "allowed values: 'gcp', 'aws' and 'azure'")
+
+        # create terraform directory
+        if not os.path.exists(self.init_dir):
+            os.makedirs(self.init_dir)
+            module = 'github.com/dcos/{}?ref={}/{}'.format(
+                'terraform-dcos-enterprise' if config['dcos-enterprise'] else 'terraform-dcos',
+                TERRAFORM_DCOS_EE_VERSION if config['dcos-enterprise'] else TERRAFORM_DCOS_VERSION,
+                config['platform'])
+            util.run_subprocess('terraform init -from-module ' + module, self.init_dir)
+
+            with open(self.cluster_profile_path, 'w') as file:
+                for k, v in config['terraform_config'].items():
+                    file.write(k + ' = ')
+                    if type(k) is dict:
+                        file.write('<<EOF\n{}\nEOF\n'.format(yaml.dump(v)))
+                    else:
+                        file.write('"{}"\n'.format(v))
+
+    def create(self):
+        # TODO ssh key helpers
+        util.run_subprocess('terraform apply -auto-approve -var-file ' + self.cluster_profile_path, self.init_dir)
+        return self.config
+
+    def wait(self):
+        """ Nothing to do here because unlike the other launchers, create() runs and also waits for a subprocess to
+        finish instead of just sending an http request to a provider.
+        """
+        pass
+
+    def delete(self):
+        util.run_subprocess('terraform destroy -force -var-file ' + self.cluster_profile_path, self.init_dir)
+        shutil.rmtree(self.init_dir, ignore_errors=True)
+
+    def describe(self) -> str:
```

Comment:
Returning exactly the terraform output here instead of parsing it with regular expressions and dumping the results in collections because I see no added value in doing so.

Reply:
A bunch of tooling will break if this isn't parsed and categorized. This is one of the major advantages of dcos-launch; you can programmatically retrieve the IPs of the masters, agents, and public agents. If you put it into a blob that is inconsistent from the other 3 deployment methods, then I don't see the point of putting this in dcos-launch because it completely breaks the consistency the project is supposed to supply

Reply:
Parsing this doesnt need to be a crazy regex. You know what fields to expect and can appropriately relabel them. I.E. go line-by-line through this output looking for a key (e.g. `Mesos Agent Public IP`) and then you know everything until the next `]` will be an IP-per-line